### PR TITLE
feat: Search

### PR DESCRIPTION
**Pull request by holloway**
supersedes #11 

* Search page and filters
* Homepage latest 3 RFCs (using search API)

Implements the OpenAPI generator/client for Red.

Screenshot from 2024-12-12 14-29-00

### CODE REVIEW

**Review by jennifer-richards — COMMENTED** [review]
Looks generally good. A couple minor comments inline - the "names" of RFC docs should be "RFC 1234" with a non-breaking space. I flagged it where I noticed but don't know if I caught everywhere. We could consider bringing back the name field to the API response schema and avoid peppering that formatting rule throughout the front end. Not sure how that balances against including effectively redundant information in the schema.

Holding off on actually approving to get @rjsparks feedback about whether we want to move forward with this or wait until we progress a bit further on returning authoritative author info in the Red API. (My own preference would be to merge this and the datatracker side of the API so that we can move forward - I think we all understand that it's still under development)

**Comment by rjsparks** [comment]
@ghwood, @alexisannerossi - please note the discussion about the presentation of RFC numbers

**Comment by holloway** [comment]
Done

**Comment by holloway** [comment]
Ready for review again now

**Review by jennifer-richards — CHANGES_REQUESTED** [review]
I'm a little surprised by the approach to converting an RFC num to a title by first constructing a "wrong" format, parsing it, and converting it to the "right" one. I suppose it has the benefit that it'll continue to work if we have other sources of RFC titles that aren't as convenient as the bare number and can deal with things that are STD / BCP.

I don't think we need to be counting CPU cycles and we'll need to refactor things when we start talking about subseries (bcp, std, etc), so not opposed. As noted inline, I think it'll misbehave if handed a name like "RFC 1234" that already has the space, so I think that should be fixed.

Also a question about `\xa0`

File: client/components/rfc.ts
Comment (holloway) [reply client/components/rfc.ts:9]: Unfortunately not, as the `h()` function escapes strings...

Screenshot from 2024-12-11 12-06-26

...and afaik HTML entities are only possible using an `innerHTML` which we probably don't want

File: client/components/rfc.ts
Comment (holloway) [reply client/components/rfc.ts:57]: Correct, it's not able to parse the formatted version (in plaintext or `h()` aka `VNode`). I could add support for plaintext parsing though.

**Comment by holloway** [comment]
@jennifer-richards 

>  other sources of RFC titles that aren't as convenient as the bare number and can deal with things that are STD / BCP.

Yep that was my thinking for paths like https://rfc-editor.org/info/std96

**Comment by holloway** [comment]
@jennifer-richards does the API provide the abstract?

**Comment by jennifer-richards** [comment]
> Should the search engine return results for `std*` like that?

I suspect the answer will be yes but we'll see. We'll need to think about what the api response should look like in those cases.

**Comment by jennifer-richards** [comment]
> @jennifer-richards does the API provide the abstract?

Yes, I believe so. I added it not long before the PR was merged so you may have grabbed a version before it did so.

File: client/components/rfc.ts
Comment (jennifer-richards) [reply client/components/rfc.ts:57]: Ok. I guess I wonder whether a more explicit construction of the `parts` array when you know the components would be cleaner than building, then parsing a string. I'll yield to your judgement about what better fits the project and what methods will be more useful, though.

**Comment by holloway** [comment]
@jennifer-richards I've added the 'abstract' now. `parseRfcId` now handles whitespace (including non-breaking spaces) and there's some unit tests to verify that

**Review by jennifer-richards — DISMISSED** [review]
LGTM

**Review by jennifer-richards — APPROVED** [review]
LGerTM

**Pull request merged**
